internal/pkg/agent/cmd/container_init_linux.go

```diff
@@ -0,0 +1,283 @@
+		if err != nil {
+			return fmt.Errorf("get cap from effective failed: %w", err)
+		}
+		if !inheritableHasCap {
+			err = procCaps.SetFlag(cap.Inheritable, true, val)
+			if err != nil {
+				return fmt.Errorf("set cap to inheritable failed: %w", err)
+			}
+			setProc = true
+		}
+	}
+
+	if !setProc {
+		return nil
+	}
+
+	if err := procCaps.SetProc(); err != nil {
+		return fmt.Errorf("set proc failed: %w", err)
+	}
+
+	return nil
+}
+
+// raiseAmbientCapabilities raises all capabilities present in the Effective set of the current process
+// to the Ambient set excluding CAP_SETPCAP, and CAP_SETFCAP.
+func raiseAmbientCapabilities() error {
+	var caps []cap.Value
+	procCaps := capProcFunc()
+
+	effectiveHasSETPCAP := false
+
+	for capVal := cap.Value(0); capVal < cap.MaxBits(); capVal++ {
+
+		exists, err := procCaps.GetFlag(cap.Effective, capVal)
+		if err != nil {
+			return fmt.Errorf("failed to get proc effective flag: %w", err)
+		}
+
+		if !exists {
+			continue
+		}
+
+		if capVal == cap.SETPCAP {
+			effectiveHasSETPCAP = true
+		}
+
+		switch capVal {
+		case cap.SETPCAP, cap.SETFCAP:
+			// don't set these as they shouldn't be required by any spawned child process
+			continue
+		default:
+		}
+
+		caps = append(caps, capVal)
+	}
+
+	if len(caps) == 0 {
+		return nil
+	}
+
+	if !effectiveHasSETPCAP {
+		return errors.New("failed to set ambient vector: missing SETPCAP capability")
+	}
+
+	iab := capBoundFunc()
+	for _, capVal := range caps {
+		if err := iab.SetVector(cap.Amb, true, capVal); err != nil {
+			return fmt.Errorf("failed to set ambient vector: %w", err)
+		}
+
+		if err := iab.SetVector(cap.Inh, true, capVal); err != nil {
+			return fmt.Errorf("failed to set ambient vector: %w", err)
+		}
+	}
+
+	return iab.SetProc()
+}
+
+// chownPaths will chown all agent related paths to the current uid and gid.
+func chownPaths(agentBaseDirectory string) error {
+	uid := os.Getuid()
+	gid := os.Getgid()
+
+	procCaps := capProcFunc()
+	hasChown, err := procCaps.GetFlag(cap.Effective, cap.CHOWN)
+	if err != nil {
+		return fmt.Errorf("failed to get chown flag: %w", err)
+	}
+	if !hasChown {
+		hasDacOverride, err := procCaps.GetFlag(cap.Effective, cap.DAC_OVERRIDE)
+		if err != nil {
+			return fmt.Errorf("failed to get dac_override flag: %w", err)
+		}
+		if !hasDacOverride {
+			return errors.New("cannot chown agent paths without CAP_CHOWN or CAP_DAC_OVERRIDE capabilities")
+		}
+	}
+
+	pathsToChown := distinctPaths{
+		agentBaseDirectory: {},
+	}
+
+	pathsToChown.addPath(envWithDefault("", "LOGS_PATH"))
+	pathsToChown.addPath(envWithDefault("", "STATE_PATH"))
+	pathsToChown.addPath(envWithDefault("", "CONFIG_PATH"))
+	pathsToChown.addPath(envWithDefault("", "DATA_PATH"))
+	pathsToChown.addPath(envWithDefault("", "HOME_PATH"))
+	return pathsToChown.chown(uid, gid)
+}
+
+// distinctPaths represents a set of paths that do not overlap.
+type distinctPaths map[string]struct{}
```

Comment:
There's no way deal with this in the standard library? https://pkg.go.dev/path/filepath#Rel perhaps?

Reply:
so my end goal here is with `distinctPaths` is to detect and exclude the overlapping paths (child->parent relation) from the recursive chowning. In a simple example 

I have these two paths:
- /path1 (agent-base)
- /path1/path2 (agent-state)

I shouldn't chown `/path1/path2` as this is gonna be chowed during `/path1` either way.
Sure `https://pkg.go.dev/path/filepath#Rel` can be used in a way to identify this parent->children path relationship but it isn't like the most straightforward way to do so 🙂 So, I don't see any great gains from switching to that but please if you tell me if have another idea I might miss something